_posts/2020-08-11-relay-pagingation-containers-demystified.markdown

```diff
@@ -0,0 +1,234 @@
+    fragment Artist_artist on Artist { # Relay has strong naming conventions
+      name # Get all the data we want here
+    }
+  `
+})
+```
+
+(At Artsy, we use [TypeScript with Relay](https://github.com/relay-tools/relay-compiler-language-typescript), but
+for this blog post we'll stick to JavaScript.)
+
+So we have a plain React component that gets some props, and a Relay fragment container that wraps it, defining the
+data that the component needs.
+
+There are other types of Relay containers beyond simple fragment containers.
+[Refetch containers](https://relay.dev/docs/en/refetch-container) are like fragment containers except you can
+refetch their contents from your GraphQL server (in response to, for example, user interaction). Using a refetch
+container is very similar to using a plain fragment container. But today, we want to talk about
+[pagination containers](https://relay.dev/docs/en/pagination-container), which use a GraphQL construct called
+_connections_ to show page after page of data.
+
+[GraphQL connections](https://www.apollographql.com/blog/explaining-graphql-connections-c48b7c3d6976/) are beyond
+the scope of this blog post, but they are a way to fetch lists of data without running into the limitations of
+returning a simple array. Connections can return metadata about their results, like how many total results there
+are and uses cursors (rather than page numbers) for paginating. They also handle when items are inserted or deleted
+from the results between requests for pages. They're pretty cool!
+
+Pagination containers take considerably more setup than plain fragment containers, and the setup itself is very
+fickle. Things simply will not work until you get the configuration _exactly correct_, and then everything works
+perfectly. The setup is largely repeated boilerplate, and what I've noticed (from other engineers but also myself)
+is that the boilerplate for new pagination containers gets copy-and-pasted from existing ones. We will see how this
+leads to small problems getting propagated throughout the codebase, and leads to engineers not feeling confident
+when working in pagination containers.
+
+So let's modify the Relay container above to fetch a list of the artist's artworks. This is a very simple example,
+only used to illustrate how to use pagination containers.
+
+```js
+class Artist extends React.Component {
+  render() {
+    return (
+    <Text>The artist name is ${this.props.artist.name}.</Text>
+    {this.props.artist.artworks.edges.map(node =>
+      /* Render each artwork */
+      <Text key={node.id}>{node.name}</Text>
+    )}
+    {this.props.relay.hasMore &&
+      <Button onPress={() => this.props.relay.loadMore() } text="Load next page" />
+    )
+  }
+}
+
+export ArtistFragmentContainer = createPaginationContainer(Artist, {
+  artist: graphql`
+    fragment Artist_artist on Artist @argumentDefinitions(
+      count: { type: "Int", defaultValue: 10 }
+      cursor: { type: "String" } {
```

Comment:
```suggestion
    fragment Artist_artist on Artist
      @argumentDefinitions(
        count: { type: "Int", defaultValue: 10 }
        cursor: { type: "String" }
      ) {
```
Maybe this is a better layout for this? The closing `)` was missing.
```
    fragment Artist_artist on Artist @argumentDefinitions(
      count: { type: "Int", defaultValue: 10 }
      cursor: { type: "String" }) {
```
This might look a bit weird.